src/psygnal/_signal.py

```diff
@@ -118,6 +121,10 @@ class attribute that is bound to the signal will be used. default None
           calling all callbacks with the arguments of the nested emitter before
           returning to continue the original emission loop.  This is the default
           behavior.
+        - "immediate-drop": Nested emission events immediately begin a deeper emission
+            loop, and the original emission loop is dis-continued.  This can be useful
+            when callbacks update their own state in response to their own emissions,
+            and the original emission loop is no longer relevant.
```

Comment:
I gotta say I don't really grok any of these, and the test is too artificial to help. Some documentation here with actual nested evented models would help here.

Anyway, as to the question of the name, I'm not sure, maybe "replace"? "immediate-drop" is certainly not intuitive as fresh-but-tired-eyes.

Reply:
thanks for weighing in.  this has been a bit of an ongoing conversation, so the best way to understand this is definitely the test (i don't think it's too artificial if you look closer) the appending to `a` is just to make assertions easier, but lets use prints instead:

```python
from psygnal import SignalInstance

# a signal that emits an integer
sig = SignalInstance((int,), recursion_mode="deferred")


def cb1(value: int) -> None:
    print(f"calling cb1 with: {value}")
    if value == 1:
        # cb1 ALSO triggers an emission of the value 2
        sig.emit(2)

def cb2(value: int) -> None:
    print(f"calling cb2 with: {value}")
    if value == 2:
        # cb2 ALSO triggers an emission of the value 3
        sig.emit(3)

def cb3(value: int) -> None:
    print(f"calling cb3 with: {value}")


sig.connect(cb1)
sig.connect(cb2)
sig.connect(cb3)
sig.emit(1)
```

with `recursion_mode="deferred"` above: you see a breadth first pattern:
ALL callbacks are called with the first emitted value, before ANY of them are called with the second emitted value (emitted by the first connected callback  `cb1`)

```
calling cb1 with: 1
calling cb2 with: 1
calling cb3 with: 1
calling cb1 with: 2
calling cb2 with: 2
calling cb3 with: 2
calling cb1 with: 3
calling cb2 with: 3
calling cb3 with: 3
```


if you change the first line to `sig = SignalInstance((int,), recursion_mode="immediate")` you see a depth first pattern:
signal emission from within any of the callbacks *immediately* goes into the next deeper nested loop of emission events, **before returning back to the original loop to call the later callbacks with the original value**


```
calling cb1 with: 1
calling cb1 with: 2
calling cb2 with: 2
calling cb1 with: 3
calling cb2 with: 3
calling cb3 with: 3
calling cb3 with: 2
calling cb2 with: 1
calling cb3 with: 1
```

With `recursion_mode="immediate-drop"`, you're simply saying: "if, during the emission loop, a callback also emits the same signal, don't bother calling any later callbacks in this loop with that original value.  **drop** them"

```
calling cb1 with: 1
calling cb1 with: 2
calling cb2 with: 2
calling cb1 with: 3
calling cb2 with: 3
calling cb3 with: 3
```


this is useful in EventedModel because that original value (`1`) is now "stale" after cb1 emits 2, and cb2 emits 3.

Reply:
i think the concept is straightforward enough, but the naming is indeed hard.  i called it "immediate-drop" because its the same as immediate (you immediately begin calling callbacks) but with the addition of "dropping" later callbacks in the outer loop.

please do give it a little thought, (i.e. speak now or forever hold your peace :joy:)

Reply:
Maybe this explanation should be putted in module level docstring?

Reply:
can do.  just thinking aloud of some other names:
- `immediate-exit` (as in, immediately enter the deeper nested loop, and exit the current loop)
- `immediate-terminal` (immediately and terminally enter the nested loop)
- `terminal` (same as above: nested/recursive emissions are "terminal" ... you never come back from them)

i guess i kinda like `terminal`?

Reply:
Maybe "drop-previous-event"?

Reply:
yep, I could go with that one too.  Will give @jni one more day to read the above

Reply:
> i don't think it's too artificial if you look closer

What? It's *absurdly* artificial! One would never write such a pattern. You've got (effectively) globals everywhere (`a` and `sig`) and callbacks directly calling the *same* signal with a different value. It is honestly extremely alien to me.

> you see a depth first pattern

I actually get the implications of the different modes on the traversal order. I just don't get in what situations I would ever use them or care about them. I have a vague idea that this is somehow related to nested evented model attributes in evented models (?), but those have in the past been dealt with (ineffectively, I'll admit) with blocking.

I presume that there is a real code problem that prompted you to add this feature, and that it did not involve appending to a global list — can you describe the problem that you are solving? imho that would be the best way to come up with a good name here.

Reply:
Thanks! That's indeed much clearer. I can see how you could end up in this situation, *and* how this mode actually makes the most sense. 🙏 

Ok now I need to sleep on it. 😂 I still find replace (you're replacing the emit loop with the new one) more intuitive than immediate-drop.

I think maybe "recursion" is the wrong way to think about it. It's kinda leaking low level details. An alternate way to think about it (not happy with exact words yet): "on_re_emission:" "enqueue" "emit_anyway" "clear".

(I'm now struggling to come up with a situation where "emit_anyway" is the right behaviour?)

Reply:
great!  

> I still find replace (you're replacing the emit loop with the new one) more intuitive than immediate-drop.

ok, let me ponder that as well.  I think my current favorite might actually be `recursion_mode='terminal'` (i.e. if you recurse, you never come back from it).    but I can see where you're going with replace

> I think maybe "recursion" is the wrong way to think about it. It's kinda leaking low level details.

also an interesting point

Reply:
because this is annoying me :) i put it to chat GPT.... who suggested the parameter `emission_strategy` with values:

- `sequential`: This value makes the system process all signals at the current level before moving to signals emitted by callbacks. This strategy ensures that every callback is called with every signal in the order the signals were emitted, but at their respective levels of emission depth.
- `interleaved`: This value allows callbacks to immediately emit and process new signals, intermixing their emissions with the processing of previously emitted signals. However, it retains some form of return to earlier emissions, creating a more intertwined execution order.
- `depth-first` or `no-backtrack`: Here, once a callback emits a new signal, the system focuses on processing this and any subsequently emitted signals, without returning to process any remaining signals at the previous level. This strategy suggests a depth-first execution without backtracking.

do any of things spark joy? :)

Reply:
ok folks.  I've changed it from `recursion_model` to `emission_strategy`, and this is now the docstring.  Will wait for @jni to wake up tonight... but then merging.  cause it's starting to feel like bike shedding on a feature that no one will probably ever change :joy:

```
    emission_strategy : Literal['nested', 'sequential', 'depth-first'] | None
        Determines the order and manner in which connected callbacks are invoked when a
        callback re-emits a signal. Default is `"nested"`.

        * `"nested"`: Signals emitted by callbacks are immediately processed in
          a deeper emission loop, before returning to process signals emitted at the
          current level.

        * `"sequential"`: Ensures *all* connected callbacks are called with the first
          emitted value, before *any* of them are called with values emitted while
          calling callbacks. Signals emitted by callbacks are enqueued for emission
          after the current level of emission is complete.

        * `"depth-first"`: This is similar to `"nested"`, but control never returns to
          the previous level of emission. When a callback emits a signal, a deeper
          level of emission is started, and the current level is discontinued.
          This strategy follows a depth-first execution without backtracking.
```

Reply:
I like emission_strategy 🙏 (maybe reemission_strategy, but that's more awkward to type so could go either way), but I dislike the value names; particularly `depth-first`, because it implies `breadth-second` which is not the case here. 😂 I would suggest:

- nested -> immediate / in_order
- sequential -> enqueue(d) / queue(d)
- depth-first -> replace / discard / clear / most_recent_only / latest_only

Thanks for your patience. 😊

Reply:
(Another potential name for the parameter: `reemission='enqueued'`, `reemission='most_recent_only'` — I kinda like it.)

Reply:
Thanks!  I will pick from those new value options and change it 👍

Reply:
I think I like reemission too!